perllib/FixMyStreet/App/Controller/Admin/Roles.pm

```diff
@@ -0,0 +1,96 @@
+package FixMyStreet::App::Controller::Admin::Roles;
+use Moose;
+use namespace::autoclean;
+
+BEGIN { extends 'Catalyst::Controller'; }
+
+use FixMyStreet::App::Form::Role;
+
+sub auto :Private {
+    my ($self, $c) = @_;
+
+    my $user = $c->user;
+    if ($user->is_superuser) {
+        $c->stash(rs => $c->model('DB::Role')->search_rs({}, { join => 'body', order_by => ['body.name', 'me.name'] }));
+    } elsif ($user->from_body) {
+        $c->stash(rs => $user->from_body->roles->search_rs({}, { order_by => 'name' }));
+    }
+}
+
+sub index :Path :Args(0) {
+    my ($self, $c) = @_;
+
+    my $p = $c->cobrand->available_permissions;
+    my %labels;
+    foreach my $group (sort keys %$p) {
+        my $group_vals = $p->{$group};
+        foreach (sort keys %$group_vals) {
+            $labels{$_} = $group_vals->{$_};
+        }
+    }
+
+    $c->stash(
+        roles => [ $c->stash->{rs}->all ],
+        labels => \%labels,
+    );
+}
+
+sub create :Local :Args(0) {
+    my ($self, $c, $id) = @_;
+
+    my $role = $c->stash->{rs}->new_result({});
+    return $self->form($c, $role);
+}
+
+sub item :PathPart('admin/roles') :Chained :CaptureArgs(1) {
+    my ($self, $c, $id) = @_;
+
+    my $obj = $c->stash->{rs}->find($id)
+        or $c->detach('/page_error_404_not_found', []);
+    $c->stash(obj => $obj);
+}
+
+sub edit :PathPart('') :Chained('item') :Args(0) {
+    my ($self, $c) = @_;
+    return $self->form($c, $c->stash->{obj});
+}
+
+sub delete :PathPart('delete') :Chained('item') :Args(0) {
+    my ($self, $c) = @_;
+    $c->stash->{obj}->delete;
```

Comment:
This line crashes if there are users in the role: `ERROR:  update or delete on table "roles" violates foreign key constraint "user_roles_role_id_fkey" on table "user_roles"`